Fix crash MXP playing MSP

vadi2 opened the pull request:
<!-- Keep the title short & concise so your mom can understand it -->
#### Brief overview of PR changes/additions
Fix crash MXP playing MSP, happened to a player in Discworld. Fixes two problems:
* retrieving a value of a tag when there was no tag
* not initialising QStrings, and later on comparing nullptr to QString which is invalid (by way of `if (mediaData.getMediaFileName() == "Off"` where getMediaFileName() returned nullptr instead of an empty QString)
#### Motivation for adding to Mudlet
Mudlet should not crash
#### Other info (issues closed, discussion etc)
Can be replicated by logging into a game with MXP and then running this:

```lua
lua feedTriggers("[4z<music Off>[4z</music>You skilfully play an upbeat version of Three Blind Mice on your wooden r")
```

vadi2 committed:
Intialise QStrings to a default value

vadi2 committed:
Don't assume FName always exists

Prevents crash when just doing this out of the blue:

lua feedTriggers("[4z<music Off>[4z</music>You skilfully play an upbeat version of Three Blind Mice on your wooden r")

vadi2 commented:
@mpconley  can you help review?

vadi2 commented:
Interesting that static analysis did not pick this up, one would assume this should be deductible?

SlySven commented on src/TMedia.h:
```diff
@@ -83,15 +83,15 @@ class TMediaData
     int mMediaProtocol;
     int mMediaType;
-    QString mMediaFileName;
+    QString mMediaFileName {};
```
But aren't these initialised to be `null` `QString`s **even *without*** an initialiser here (or in the constructor initialiser list)?

vadi2 replied:
I don't know the exact details - could be `null` and that you can't compare `null` one to QString, could be something else, but it fixes the problem.

SlySven replied:
Fundamentally I do not think this - and the changes on line 91 to 94 are making the slightest difference! The default constructor for a `QString` is used automagically and putting a `{}` as an initialiser here (or including the `QString` member in the constructor's list with a default `()` argument) has precisely the same effect as not explicitly initialising them at all...

🤓 

OTOH Comparing a `QString` instance to a `nullptr` ***IS*** wrong...

mpconley commented on src/TMxpMusicTagHandler.cpp:
```diff
@@ -29,7 +29,9 @@ TMxpTagHandlerResult TMxpMusicTagHandler::handleStartTag(TMxpContext& ctx, TMxpC
     mediaData.setMediaProtocol(TMediaData::MediaProtocolMSP);
     mediaData.setMediaType(TMediaData::MediaTypeMusic);
-    mediaData.setMediaFileName(tag->getAttributeValue("FName"));
+    if (tag->hasAttribute("FName")) {
```
Every MXP MSP tag should have an FName (which is either `Off` or a file name). I recall elsewhere in the code where there was some validation. I wonder if adding this here may do more harm than good - or perhaps have no effect due to validation elsewhere. 

If this change is appropriate here, it may also need added to TMxpSoundTagHandler::handleStartTag as well.

SlySven replied:
This could be the actual bugfix...

vadi2 replied:
This was necessary in combination with the other bugfix during my development. If you're not sold feel free to replicate the scenario!

SlySven replied:
I do not have the confidence/ability/any MUDS setup to frob with MXP stuff - but I feel sure that removing the explicit initialisations of those `QString` members on lines 86 and 91 to 94 will not cause the bug to return.

mpconley commented on src/TMedia.h:
```diff
@@ -83,15 +83,15 @@ class TMediaData
+    QString mMediaTag {};
+    QString mMediaUrl {};
+    QString mMediaKey {};
+    QString mMediaAbsolutePathFileName {};
```
Give me a green light and I will dig up my Windows machine to test this (as we know my new Macbook is not taking kindly to the latest builds).

vadi2 replied:
@mpconley mac builds back online, links are updated.

mpconley replied:
Thanks @vadi2 I'll plan to run some music tests by the end of this weekend.

SlySven commented on src/TMxpMusicTagHandler.cpp:
```diff
@@ -29,7 +29,9 @@ TMxpTagHandlerResult TMxpMusicTagHandler::handleStartTag(TMxpContext& ctx, TMxpC
-    mediaData.setMediaFileName(tag->getAttributeValue("FName"));
+    if (tag->hasAttribute("FName")) {
+        mediaData.setMediaFileName(tag->getAttributeValue("FName"));
+    }
```
If https://github.com/Mudlet/Mudlet/pull/4577/files#r551074385 is the case then should there not be a further change here:
```suggestion
    } else {
        mediaData.setMediaFileName(QStringLiteral("off"));
    }
```
?

vadi2 committed:
Merge remote-tracking branch 'origin/development' into fix-crash-mxp-msp

vadi2 committed:
Merge remote-tracking branch 'origin/development' into fix-crash-mxp-msp

mpconley commented:
Perhaps efforts should be directed at investigating TMxpTagProcessor.cpp and supporting code, first.  The same problem that exists for **both** the Sound and Music tags may also exist for all MXP tags.

**Pre-Read**
Checkout the requirements here: [http://www.zuggsoft.com/zmud/mxp.htm#ATTLIST](http://www.zuggsoft.com/zmud/mxp.htm#ATTLIST) for how MXP attributes/values are intended to be handled.  They use a Sound tag as an example.

**How it Should Be**
The tag definitions are found inside of [TMxpTagProcessor.cpp](https://github.com/Mudlet/Mudlet/blob/development/src/TMxpTagProcessor.cpp).  Per the MXP spec, one could send an attribute/value pair like `fname=Off` into the tags or skip sending the attribute and only send the value like `Off`.  If only values are sent, tag processing should use the order of the information sent and has options for assuming default values for missing attribute/value pairs and sending that on to helper classes for more processing.

A Sound or Music tag alone with no attributes or values is invalid.  There should always be at least one value, like `<sound Off>` or `<music ambience.mp3>` and better yet `<sound fname=Off>` or `<music fname=ambience.mp3>`.

**Assumptions / Poor Testing**
When the Sound and Music tags were created, it was assumed that the "no attribute" part of the MXP spec was part of the TMxpTagProcessor.cpp and its supporting classes, but the author (me) may have not tested all aspects.  Given the example sent in `<music Off>` it does not appear the processor is applying the `Off` to the fname attribute and sending it into the Sound and Music tag handlers.

**Next steps**
The TMxpTagProcessor.cpp may require a good look to ensure it is following the spec for all MXP tags and whether the author of the framework and tags other than Sound and Music pushed logic further into the individual tags.  If the logic is deeper in the tags, add similar logic for Sound and Music to account for the variety of attribute/value pairs sent in.  

**To Be Considered**
This PR, or more PRs before the next release, need to handle **both** the Sound and Music tags and not just Music.  The team may also need to investigate if the condition exists for the remainder of the MXP tags.

**Cautions**
Making spot changes to TMedia code could have unintended consequences for general sound and music handling outside of MXP, such as MSP and MCMP code.  Although making the interior of TMedia more robust may be a good decision now, there is some chance it could destabilize MSP and MCMP.

The [MSP spec](http://www.zuggsoft.com/zmud/msp.htm) was also very specific that fName=Off (with a capital `O`).  So far, that is consistent in the code base.  We could make that case-insensitive, but we should be consistent.

vadi2 commented:
We should do that but I haven't got the bandwidth for that right now, and we have a game that can crash Mudlet on our hands first - I think the bandaid is worthwhile to be applied.

vadi2 committed:
Merge remote-tracking branch 'vadi2/fix-crash-mxp-msp' into fix-crash-mxp-msp

mpconley commented:
I get a crash unfortunately on the Mac with M1 chip. It is only happening on StickMUD, but StickMUD is the only game I am connecting to that supports sound. I try to connect, then it asks me for my Mac keychain password, then I have seen some evidence a connection was made and negotiation was running, then there is the crash.

[Mudlet_Crash_02132021.txt](https://github.com/Mudlet/Mudlet/files/5976009/Mudlet_Crash_02132021.txt)

vadi2 committed:
Merge remote-tracking branch 'origin/development' into fix-crash-mxp-msp

vadi2 commented:
I love how good macOS is about giving complete stacktraces. If only every OS did that.

In this case it's a recent issue that we fixed (https://github.com/Mudlet/Mudlet/issues/4776), sorry about that. Updated this branch with the fix - it'll work well once links for 4eddc59 are built.

mpconley commented:
I did some testing on Mac.  I recommend a different approach.  The original issue is not satisfied as-is.  It could solve the crasher, but the tag will be useless.  There is a fairly simple design gap with TMxpMusicTagHandler.cpp and TMxpSoundTagHandler.cpp, and thankfully there is a pattern to follow inside TMxpSendTagHandler.cpp to cure the gap and solve the issue.

The Sound and Music MXP tags were not designed or tested to handle a part of the MXP specification where attribute names are left out of the syntax. For example, `<music FName="Off">` works great now sending the `Fname` attribute name, but `<music "Off">` or `<music Off>` do not work. The pattern used inside TMxpSendTagHandler.cpp around methods containing the name `extract` (and methods using those methods) needs applied in the TMxpMusicTagHandler.cpp and TMxpSoundTagHandler.cpp files to support not passing in attribute names.

@vadi2 you'd need to look at how attributes are listed in TMxpTagProcessor.cpp, seeing what processing was done with the `extract` methods in TMxpSendTagHandler.cpp, then looking again at TMxpTagProcessor.cpp for Sound and Music and doing the same for TMxpMusicTagHandler.cpp and TMxpSoundTagHandler.cpp.

I could take on that task, but I'll need to see if I could get my new device compiling Mudlet again, which may take time. 

During testing these new issues were discovered:

- With MSP over MXP, passing in a "U" value with a URL for Sound and Music MXP tags is not downloading or playing the file and it freezes the UI when MXP is used in StickMUD at least.

- Outside of MXP, with MCMP, there is a problem with "priority" management. The problem does not exist with the MSP, MXP, or telopt flavors of the TMedia Sound and Music code. This had to have worked in past releases. It needs a look. Likely another issue.

vadi2 commented:
If you could give it a go that would be good. I'm not all that across the new MXP code.

Kebap commented:
> I could take on that task, but I'll need to see if I could get my new device compiling Mudlet again, which may take time.

Maybe the automatic builds in a (draft) PR can be of assistance meanwhile? 
Then again, some devs recommend our github codespaces for compiling as well.

mpconley commented:
Good ideas.  I would like to get this fix in the next deployment also.

On Wed, Feb 17, 2021 at 7:04 AM Kebap <notifications@github.com> wrote:

> I could take on that task, but I'll need to see if I could get my new
> device compiling Mudlet again, which may take time.
>
> Maybe the automatic builds in a (draft) PR can be of assistance meanwhile?
> Then again, some devs recommend our github codespaces for compiling as
> well.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Mudlet/Mudlet/pull/4577#issuecomment-780511297>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAXKUARSCHN67HGMFY4BF3DS7OWFPANCNFSM4VRXXHSA>
> .
>

vadi2 commented:
Closing in favour of https://github.com/Mudlet/Mudlet/pull/4839
